src/core/model/StrokeContour.cpp

```diff
@@ -0,0 +1,323 @@
+#include "StrokeContour.h"
+
+#include <cmath>
+
+#include "model/MathVect.h"
+#include "model/Point.h"
+#include "util/Assert.h"
+
+static_assert(std::numeric_limits<double>::is_iec559);  // Ensures atan2(0., 0.) does not error
+
+/// Takes a double between -2*pi and 2*pi and returns the corresponding angle between -pi and pi
+static constexpr double inMinusPiPiInterval(double t) {
+    xoj_assert(t >= -2 * M_PI && t <= 2 * M_PI);
+    return t > M_PI ? t - 2 * M_PI : t < -M_PI ? t + 2 * M_PI : t;
+}
+
+// Adds to cairo a segment corresponding to the cord of an arc obtained via cairo_arc() (with the same parameters)
+static void drawChord(cairo_t* cr, double cx, double cy, double radius, double angle1, double angle2) {
+    cairo_line_to(cr, cx + radius * std::cos(angle1), cy + radius * std::sin(angle1));
+    cairo_line_to(cr, cx + radius * std::cos(angle2), cy + radius * std::sin(angle2));
+}
+
+static void hookAfter(cairo_t* cr, double x, double y, double r, double a, double b) {
+    cairo_line_to(cr, x, y);
+    cairo_arc(cr, x, y, r, a, b);
+}
+static void hookBefore(cairo_t* cr, double x, double y, double r, double a, double b) {
+    cairo_arc(cr, x, y, r, a, b);
+    cairo_line_to(cr, x, y);
+}
+
+struct ReturnOp {
+    ReturnOp(void (*op)(cairo_t*, double, double, double, double, double), double x, double y, double r, double a,
+             double b):
+            op(op), x(x), y(y), r(r), a(a), b(b) {}
+    void (*op)(cairo_t*, double, double, double, double, double);
+    double x, y, r, a, b;
+
+    void operator()(cairo_t* cr) { op(cr, x, y, r, a, b); }
+};
+
+xoj::view::StrokeContour::StrokeContour(const std::vector<Point>& path): path(path) {}
+xoj::view::StrokeContour::~StrokeContour() = default;
+
+static inline void drawCoupling(cairo_t* cr, std::vector<ReturnOp>& ops, const Point& p2, double n1, double n3,
+                                double a1, double a3, double z1) {
+    if (z1 < p2.z) {
+        if (.5 * p2.z < n1  // Optimisation to avoid computing the std::hypot
+            || .5 * p2.z <= std::hypot(.5 * z1, n1)) [[likely]] {
+            double angleIn = std::asin(z1 / p2.z);
+            double a = inMinusPiPiInterval(a1 + angleIn);
+            double b = inMinusPiPiInterval(a3 - M_PI_2);
+            auto* arcFun =
+                    (inMinusPiPiInterval(a3 - a1) <= 0.0 || inMinusPiPiInterval(b - a) > 0.0) ? cairo_arc : drawChord;
+            arcFun(cr, p2.x, p2.y, .5 * p2.z, a, b);
+
+            double bp = inMinusPiPiInterval(a1 - angleIn);
+            double ap = inMinusPiPiInterval(a3 + M_PI_2);
+            auto* arcFunp =
+                    (inMinusPiPiInterval(a3 - a1) >= 0.0 || inMinusPiPiInterval(bp - ap) > 0.0) ? cairo_arc : drawChord;
+            ops.emplace_back(arcFunp, p2.x, p2.y, .5 * p2.z, ap, bp);
+        } else {
+            cairo_line_to(cr, p2.x, p2.y);
+            cairo_arc(cr, p2.x, p2.y, .5 * p2.z, a1, a3 - M_PI_2);
```

Comment:
Can you please explain this important case? I see where the path must head to, but why in this precise way?